File: mac-app/QuartzTeachers/Scripting/TaskMilestones.swift
import Foundation

/// One step of a long-running task: the text shown to the teacher, and
/// the marker in the output that means the step has been reached.
struct TaskMilestone {

    // MARK: - Stored properties

    /// Brief, meaningful label, e.g. "Building your site".
    let label: String

    /// A phrase that appears in the output when this step begins.
    let marker: String
}

/// The ordered milestones for each kind of task. Progress is "how many
/// milestones have been reached", so the bar advances in real steps
/// rather than spinning indefinitely.
enum TaskMilestones {

    // MARK: - Stored properties

    /// Creating a course: the wizard's own progression.
    static let courseCreation: [TaskMilestone] = [
        TaskMilestone(label: "Building your website builder…", marker: "Building your website builder"),
        TaskMilestone(label: "Getting ready…", marker: "Welcome to the Course Setup Script"),
        TaskMilestone(label: "Preparing your course folder…", marker: "'Media' folder"),
        TaskMilestone(label: "Choosing appearance…", marker: "Quartz Locale"),
        TaskMilestone(label: "Setting up sections…", marker: "timetable section numbers"),
        TaskMilestone(label: "Creating folders and files…", marker: "Select folders/files to HIDE"),
        TaskMilestone(label: "Finishing up…", marker: "set up successfully"),
    ]

    /// Installing the example course.
    static let exampleCourse: [TaskMilestone] = [
        TaskMilestone(label: "Building your website builder…", marker: "Building your website builder"),
        TaskMilestone(label: "Getting things ready…", marker: "Starting container if needed"),
        TaskMilestone(label: "Copying the example course…", marker: "Example Course installed to"),
        TaskMilestone(label: "Finishing up…", marker: "EXAMPLE_COURSE_CODE="),
    ]

    /// Previewing a section.
    static let preview: [TaskMilestone] = [
        TaskMilestone(label: "Building your website builder…", marker: "Building your website builder"),
        TaskMilestone(label: "Starting up…", marker: "Starting container if needed"),
        TaskMilestone(label: "Gathering your content…", marker: "Copying shared folders"),
        TaskMilestone(label: "Applying your settings…", marker: "Updated pageTitle"),
        TaskMilestone(label: "Preparing components…", marker: "Installing dependencies"),
        TaskMilestone(label: "Building your site…", marker: "Quartz v4"),
        TaskMilestone(label: "Opening the preview…", marker: "Launching Quartz preview"),
    ]

    /// Publishing a section to Netlify.
    static let deploy: [TaskMilestone] = [
        TaskMilestone(label: "Building your website builder…", marker: "Building your website builder"),
        TaskMilestone(label: "Starting up…", marker: "Ensuring container is running"),
        TaskMilestone(label: "Checking your site…", marker: "Deploying from local build"),
        TaskMilestone(label: "Connecting to Netlify…", marker: "Netlify site"),
        TaskMilestone(label: "Comparing what changed…", marker: "delta deploy manifest"),
        TaskMilestone(label: "Uploading your pages…", marker: "Delta deploy created"),
        TaskMilestone(label: "Finishing up…", marker: "Deploy complete"),
    ]

    /// Publishing when the site has to be rebuilt first — one task from
    /// the teacher's point of view, so one progress bar.
    static let buildAndDeploy: [TaskMilestone] = [
        TaskMilestone(label: "Building your website builder…", marker: "Building your website builder"),
        TaskMilestone(label: "Starting up…", marker: "Starting container if needed"),
        TaskMilestone(label: "Gathering your content…", marker: "Copying shared folders"),
        TaskMilestone(label: "Building your site…", marker: "Quartz v4"),
        TaskMilestone(label: "Connecting to Netlify…", marker: "Netlify site"),
        TaskMilestone(label: "Uploading your pages…", marker: "Delta deploy created"),
        TaskMilestone(label: "Finishing up…", marker: "Deploy complete"),
    ]

}

File: windows-app/Plantoir.Core/Scripting/TaskMilestones.cs
namespace Plantoir.Core.Scripting;

/// <summary>A label shown to the teacher and the marker searched for in output.</summary>
public sealed record TaskMilestone(string Label, string Marker);

/// <summary>
/// The ordered milestone lists for each task, with markers matched against
/// what the WINDOWS launchers really print (the .ps1 output differs from the
/// .sh output in its host-side lines; the in-container Python lines are
/// shared). Every label ends with an ellipsis — the shared invariant.
/// </summary>
public static class TaskMilestones
{
    public static readonly IReadOnlyList<TaskMilestone> CourseCreation = new[]
    {
        new TaskMilestone("Getting things ready…", "Detected host timezone offset"),
        new TaskMilestone("Getting ready…", "Welcome to the Course Setup Script"),
        new TaskMilestone("Preparing your course folder…", "'Media' folder"),
        new TaskMilestone("Choosing appearance…", "Quartz Locale"),
        new TaskMilestone("Setting up sections…", "timetable section numbers"),
        new TaskMilestone("Creating folders and files…", "Select folders/files to HIDE"),
        new TaskMilestone("Finishing up…", "set up successfully"),
    };

    public static readonly IReadOnlyList<TaskMilestone> ExampleCourse = new[]
    {
        new TaskMilestone("Getting things ready…", "Detected host timezone offset"),
        new TaskMilestone("Copying the example course…", "Example Course installed to"),
        new TaskMilestone("Finishing up…", "EXAMPLE_COURSE_CODE="),
    };

    public static readonly IReadOnlyList<TaskMilestone> Preview = new[]
    {
        new TaskMilestone("Getting things ready…", "Running the website builder on this PC"),
        new TaskMilestone("Gathering your content…", "Copying shared folders"),
        new TaskMilestone("Applying your settings…", "Updated pageTitle"),
        new TaskMilestone("Preparing components…", "Installing dependencies"),
        new TaskMilestone("Building your site…", "Quartz v4"),
        new TaskMilestone("Opening the preview…", "Launching Quartz preview"),
    };

    public static readonly IReadOnlyList<TaskMilestone> Deploy = new[]
    {
        new TaskMilestone("Getting things ready…", "Host timezone offset"),
        new TaskMilestone("Starting up…", "from this PC"),
        new TaskMilestone("Checking your site…", "Deploying from local build"),
        new TaskMilestone("Connecting to Netlify…", "Netlify site"),
        new TaskMilestone("Comparing what changed…", "delta deploy manifest"),
        new TaskMilestone("Uploading your pages…", "Delta deploy created"),
        new TaskMilestone("Finishing up…", "Deploy complete"),
    };

    /// <summary>Rebuild + publish presented as ONE task with one bar.</summary>
    public static readonly IReadOnlyList<TaskMilestone> BuildAndDeploy = new[]
    {
        new TaskMilestone("Getting things ready…", "Running the website builder on this PC"),
        new TaskMilestone("Gathering your content…", "Copying shared folders"),
        new TaskMilestone("Building your site…", "Quartz v4"),
        new TaskMilestone("Connecting to Netlify…", "Netlify site"),
        new TaskMilestone("Uploading your pages…", "Delta deploy created"),
        new TaskMilestone("Finishing up…", "Deploy complete"),
    };

    /// <summary>
    /// Folder-mode publish: the site is already built on the host, so the
    /// launcher only mirrors changed files into the chosen folder — quick,
    /// containerless, and never mentioning Netlify (row 102d).
    /// </summary>
    public static readonly IReadOnlyList<TaskMilestone> DeployToFolder = new[]
    {
        new TaskMilestone("Checking your site…", "Host timezone offset"),
        new TaskMilestone("Copying your files…", "to a folder"),
        new TaskMilestone("Finishing up…", "PUBLISHED_FOLDER="),
    };

    /// <summary>Rebuild + folder publish presented as ONE task with one bar.</summary>
    public static readonly IReadOnlyList<TaskMilestone> BuildAndDeployToFolder = new[]
    {
        new TaskMilestone("Getting things ready…", "Running the website builder on this PC"),
        new TaskMilestone("Gathering your content…", "Copying shared folders"),
        new TaskMilestone("Building your site…", "Quartz v4"),
        new TaskMilestone("Copying your files…", "to a folder"),
        new TaskMilestone("Finishing up…", "PUBLISHED_FOLDER="),
    };

    /// <summary>
    /// Cloudflare publish. The phrases match what deploy.py prints on that
    /// path — and, like folder mode, never say Netlify.
    /// </summary>
    public static readonly IReadOnlyList<TaskMilestone> DeployToCloudflare = new[]
    {
        new TaskMilestone("Getting things ready…", "Host timezone offset"),
        new TaskMilestone("Starting up…", "from this PC"),
        new TaskMilestone("Checking your site…", "Deploying from local build"),
        new TaskMilestone("Connecting to Cloudflare…", "Cloudflare project"),
        new TaskMilestone("Uploading your pages…", "Uploading the built site"),
        new TaskMilestone("Finishing up…", "Deploy complete"),
    };

    /// <summary>Rebuild + Cloudflare publish presented as ONE task with one bar.</summary>
    public static readonly IReadOnlyList<TaskMilestone> BuildAndDeployToCloudflare = new[]
    {
        new TaskMilestone("Getting things ready…", "Running the website builder on this PC"),
        new TaskMilestone("Gathering your content…", "Copying shared folders"),
        new TaskMilestone("Building your site…", "Quartz v4"),
        new TaskMilestone("Connecting to Cloudflare…", "Cloudflare project"),
        new TaskMilestone("Uploading your pages…", "Uploading the built site"),
        new TaskMilestone("Finishing up…", "Deploy complete"),
    };

    public static IEnumerable<IReadOnlyList<TaskMilestone>> AllLists
    {
        get
        {
            yield return CourseCreation;
            yield return ExampleCourse;
            yield return Preview;
            yield return Deploy;
            yield return BuildAndDeploy;
            yield return DeployToFolder;
            yield return BuildAndDeployToFolder;
            yield return DeployToCloudflare;
            yield return BuildAndDeployToCloudflare;
        }
    }
}
